### ISSUE TITLE
Negative densities

### ISSUE THREAD

**Issue by jason5306** [Ruyi-Zha/r2_gaussian#24]
Hi Ruyi,
I wonder if r2 gaussian output negative densities, for example the to-be-reconstructed volume has negative intensities. In such case, does that mean I need to normalize the volume to [0,1]?

Thank you!

**Comment by Ruyi-Zha** [comment]
Hi, to support negative intensities, you may need to edit certain lines in the CUDA forward and backward functions. For instance, you could start by commenting out this line and similar ones in other forward and backward functions.
https://github.com/Ruyi-Zha/r2_gaussian/blob/788971da1a7e8a41d79bcc25c05e0e54f9dcb748/r2_gaussian/submodules/xray-gaussian-rasterization-voxelization/cuda_rasterizer/forward.cu#L374

Additionally, consider modifying the density_activation function, as it currently uses `softplus`—which only outputs values in the range of [0,+inf].

https://github.com/Ruyi-Zha/r2_gaussian/blob/788971da1a7e8a41d79bcc25c05e0e54f9dcb748/r2_gaussian/gaussian/gaussian_model.py#L61

**Comment by jason5306** [comment]
Thanks a lot, I'll try if it work. Thank you!

**Comment by jason5306** [comment]
> Hi, to support negative intensities, you may need to edit certain lines in the CUDA forward and backward functions. For instance, you could start by commenting out this line and similar ones in other functions.
> 
> [r2_gaussian/r2_gaussian/submodules/xray-gaussian-rasterization-voxelization/cuda_rasterizer/forward.cu](https://github.com/Ruyi-Zha/r2_gaussian/blob/788971da1a7e8a41d79bcc25c05e0e54f9dcb748/r2_gaussian/submodules/xray-gaussian-rasterization-voxelization/cuda_rasterizer/forward.cu#L374)
> 
> Line 374 in [788971d](/Ruyi-Zha/r2_gaussian/commit/788971da1a7e8a41d79bcc25c05e0e54f9dcb748)
> 
>  if (alpha < 0.00001f) 
> Additionally, consider modifying the density_activation function, as it currently uses `softplus`—which only outputs values in the range of [0,+inf].
> 
> [r2_gaussian/r2_gaussian/gaussian/gaussian_model.py](https://github.com/Ruyi-Zha/r2_gaussian/blob/788971da1a7e8a41d79bcc25c05e0e54f9dcb748/r2_gaussian/gaussian/gaussian_model.py#L61)
> 
> Line 61 in [788971d](/Ruyi-Zha/r2_gaussian/commit/788971da1a7e8a41d79bcc25c05e0e54f9dcb748)
> 
>  self.density_activation = torch.nn.Softplus()  # use softplus for [0, +inf]



May I ask if there's a particular reason why in r2 gaussian the activation function used is softplus while the original 3DGS was using sigmoid? Also, I wonder if the used of softplus is to prevent the activated density from zero when the raw density is very small? Thanks!

**Comment by Ruyi-Zha** [comment]
Hi, there's no particular reason for that. If I remember correctly, Softplus slightly improves stability, but I believe both activation functions work.

**Comment by jason5306** [comment]
I see, I thought there was a particular reason of using an unsaturated activation for radiative gs. Thanks!
